### PR TITLE
Use status list 2021 context

### PR DESCRIPTION
**Pull request by JSAssassin**
- Update texts, docs, and tests. 
- Add additional tests.

### CODE REVIEW

**Review by dmitrizagidulin — APPROVED** [review]
👍

File: main.js [hunk 1]
```diff
@@ -1,31 +1,31 @@
...
 import vc from '@digitalbazaar/vc';
+import statusListCtx from 'vc-status-list-context';
+import credentialsCtx from 'credentials-context';
 
-const CONTEXTS = {
-  VC_V1: 'https://www.w3.org/2018/credentials/v1',
-  RL_V1: 'https://w3id.org/vc-revocation-list-2020/v1'
-};
+const VC_V1_CONTEXT = credentialsCtx.constants.CREDENTIALS_CONTEXT_V1_URL;
+const SL_V1_CONTEXT = statusListCtx.constants.CONTEXT_URL_V1;
```
[the remark this answers is not in the document]
Comment (JSAssassin) [reply main.js]: Updated here 2b578e7

File: package.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
-  "name": "vc-revocation-list",
+  "name": "vc-status-list",
   "version": "3.0.1-0",
```
[the remark this answers is not in the document]
Comment (JSAssassin) [reply package.json]: Fixed e0edf2c

**Review by davidlehn — APPROVED** [review]

**Review by msporny — APPROVED** [review]

File: StatusList.js [hunk 3]
```diff
@@ -1,9 +1,9 @@
 /*!
- * Copyright (c) 2020-2021 Digital Bazaar, Inc. All rights reserved.
+ * Copyright (c) 2022 Digital Bazaar, Inc. All rights reserved.
  */
 import Bitstring from '@digitalbazaar/bitstring';
 
-export default class RevocationList {
+export default class StatusList {
```
Comment (dlongley) [inline StatusList.js:6]: If this is a `StatusList`, why does it have `setRevoked` and `isRevoked` in it? Isn't this just a `RevocationList` again? Should there be a base class that does `setStatus` -- and `RevocationList` inherits and calls that internally from `setRevoked`? Or would the base class just suffice -- and you need to know what you're working with?

Something here needs adjustment.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: StatusList.js (hunk 3 above)
Comment (JSAssassin) [reply StatusList.js]: I have updated RevocationList to be a base class here [7d1a7b6](https://github.com/digitalbazaar/vc-status-list/pull/1/commits/7d1a7b6b833805037cd74e1a4fa1a6a4e1b7adff) Is this what you had in mind?

File: StatusList.js [hunk 4]
```diff
@@ -19,13 +19,19 @@ export default class RevocationList {
   isRevoked(index) {
     return this.bitstring.get(index);
   }
+}
+
+export default class StatusList extends RevocationList {
```
Comment (dlongley) [inline StatusList.js:24]: This is backwards -- the StatusList should be the base class and, if we want a RevocationList class, it should extend the StatusList class and provide specific names for its status functions. Also, do not use `default` exports.

Comment (dlongley) [reply StatusList.js:24]: I recommend trying to get away with not needing subclasses for different types of statuses -- and instead just use generic names for the methods.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: StatusList.js (hunk 4 above)
Comment (JSAssassin) [reply StatusList.js]: Updated here [9363d26](https://github.com/digitalbazaar/vc-status-list/pull/1/commits/9363d26883ec5695754b8f229230ed0c723c6edd)

File: tests/20-StatusList.spec.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      list = await StatusList.decode({encodedList: encodedList100k});
+    } catch(e) {
+      err = e;
+    }
+    should.not.exist(err);
+    should.exist(list);
+    list.length.should.equal(100000);
+  });
+
+  it('should setStatus of a credential revoked', async () => {
```
Comment (dlongley) [inline tests/20-StatusList.spec.js:57]:
```suggestion
  it('should mark a credential revoked', async () => {
```

File: main.js [hunk 6]
```diff
@@ -65,8 +65,9 @@ export function statusTypeMatches({credential} = {}) {
   if(!Array.isArray(contexts)) {
     throw new TypeError('"@context" must be an array.');
   }
-  if(contexts[0] !== CONTEXTS.VC_V1) {
-    throw new Error(`The first "@context" value must be "${CONTEXTS.VC_V1}".`);
+  if(contexts[0] !== VC_V1_CONTEXT_URL) {
+    throw new Error(`The first "@context" value must be ' +
+      '"${VC_V1_CONTEXT_URL}".`);
```
Comment (dlongley) [inline main.js:70]: This line opens with a backtick and intends to close with a single quote -- but instead erroneously becomes a multiline string. Regardless, it should have been dropped to the next line:

```suggestion
    throw new Error(
      `The first "@context" value must be "${VC_V1_CONTEXT_URL}".`);
```

File: main.js [hunk 7]
```diff
@@ -97,32 +99,35 @@ export function assertRevocationList2020Context({credential} = {}) {
...
```
Comment (dlongley) [inline main.js:104]: Same problem here.

```suggestion
    throw new Error(
      `The first "@context" value must be "${VC_V1_CONTEXT_URL}".`);
```

File: main.js [hunk 8]
```diff
@@ -97,32 +99,35 @@ export function assertRevocationList2020Context({credential} = {}) {
...
     typeof credential.credentialStatus === 'object')) {
     throw new Error('"credentialStatus" is missing or invalid.');
   }
   const {credentialStatus} = credential;
-  if(credentialStatus.type !== 'RevocationList2020Status') {
+  if(!(credentialStatus.type === 'RevocationList2021Status' ||
+    credentialStatus.type === 'SuspensionList2021Status')) {
     throw new Error(
-      '"credentialStatus" type is not "RevocationList2020Status".');
+      '"credentialStatus" type must be "RevocationList2021Status" or ' +
```
Comment (dlongley) [inline main.js:125]:
```suggestion
      '"credentialStatus.type" must be "RevocationList2021Status" or ' +
```

File: main.js [hunk 9]
```diff
@@ -97,32 +99,35 @@ export function assertRevocationList2020Context({credential} = {}) {
...
+        '"SuspensionList2021Status".');
   }
-  if(typeof credentialStatus.revocationListCredential !== 'string') {
+  if(typeof credentialStatus.statusListCredential !== 'string') {
     throw new TypeError(
-      '"credentialStatus" revocationListCredential must be a string.');
+      '"credentialStatus" statusListCredential must be a string.');
```
Comment (dlongley) [inline main.js:130]:
```suggestion
      '"credentialStatus.statusListCredential" must be a string.');
```

File: main.js [hunk 10]
```diff
@@ -194,31 +198,32 @@ async function _checkStatus({
...
+      typeof slCredential.issuer === 'object' ?
+        slCredential.issuer.id : slCredential.issuer;
 
-    if(!(credentialIssuer && revocationListCredentialIssuer) ||
-      (credentialIssuer !== revocationListCredentialIssuer)) {
-      throw new Error('Issuers of the revocation credential and verifiable ' +
+    if(!(credentialIssuer && statusListCredentialIssuer) ||
+      (credentialIssuer !== statusListCredentialIssuer)) {
+      throw new Error('Issuers of the status list credential and verifiable ' +
         'credential do not match.');
```
Comment (dlongley) [inline main.js:214]:
```suggestion
      throw new Error(
        'Issuers of the status list credential and verifiable ' +
        'credential do not match.');
```

File: main.js [hunk 11]
```diff
@@ -194,31 +198,32 @@ async function _checkStatus({
...
     }
   }
-
-  if(!rlCredential.type.includes('RevocationList2020Credential')) {
-    throw new Error('"RevocationList2020Credential" type is not valid.');
+  if(!slCredential.type.includes('StatusList2021Credential')) {
+    throw new Error(
+      'Status list credential must include "StatusList2021Credential" type.');
```
Comment (dlongley) [inline main.js:219]:
```suggestion
      'Status list credential type must include "StatusList2021Credential".');
```

File: tests/20-StatusList.spec.js [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    list.getStatus(1).should.equal(false);
+    list.getStatus(2).should.equal(false);
+    list.getStatus(3).should.equal(false);
+    list.getStatus(4).should.equal(true);
+    list.getStatus(5).should.equal(false);
+    list.getStatus(6).should.equal(false);
+    list.getStatus(7).should.equal(false);
+  });
+
+  it('should fail to mark a credential revoked no "revoked" boolean param ' +
```
Comment (dlongley) [inline tests/20-StatusList.spec.js:78]: Will need to change based on new param name.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: StatusList.js [hunk 13]
```diff
@@ -1,22 +1,22 @@
...
 
-export default class RevocationList {
+export class StatusList {
   constructor({length, buffer} = {}) {
     this.bitstring = new Bitstring({length, buffer});
     this.length = this.bitstring.length;
   }
 
-  setRevoked(index, revoked) {
+  setStatus(index, revoked) {
```
[the remark this answers is not in the document]
Comment (JSAssassin) [reply StatusList.js]: Renamed the param to `revokedStatus` here [5e4f973](https://github.com/digitalbazaar/vc-status-list/pull/1/commits/5e4f973b7abafa934c5f8e3fc6289e0feb8d576c). I am not sure if that's a good name,  any suggestions?

File: StatusList.js [hunk 14]
```diff
@@ -1,22 +1,22 @@
...
+export class StatusList {
   constructor({length, buffer} = {}) {
     this.bitstring = new Bitstring({length, buffer});
     this.length = this.bitstring.length;
   }
 
-  setRevoked(index, revoked) {
-    if(typeof revoked !== 'boolean') {
-      throw new TypeError('"revoked" must be a boolean.');
+  setStatus(index, revokedStatus) {
```
Comment (dlongley) [inline StatusList.js:12]: This needs to be renamed to be more generic to support statuses other than "revocation" (such as "suspension").

```suggestion
  setStatus(index, status) {
```

File: StatusList.js [hunk 15]
```diff
@@ -1,22 +1,22 @@
...
+    if(typeof revokedStatus !== 'boolean') {
+      throw new TypeError('"revokedStatus" must be a boolean.');
```
Comment (dlongley) [inline StatusList.js:14]:
```suggestion
    if(typeof status !== 'boolean') {
      throw new TypeError('"status" must be a boolean.');
```

File: StatusList.js [hunk 16]
```diff
@@ -1,22 +1,22 @@
...
     }
-    return this.bitstring.set(index, revoked);
+    return this.bitstring.set(index, revokedStatus);
```
Comment (dlongley) [inline StatusList.js:16]:
```suggestion
    return this.bitstring.set(index, status);
```

**Review by dlongley — CHANGES_REQUESTED** [review]

**Review by dlongley — APPROVED** [review]
Thanks!

**Pull request merged**